### PR TITLE
Core/Spells: Implemented spell effect 286 (SPELL_EFFECT_GRANT_BATTLEPET_EXPERIENCE)

### PR DESCRIPTION
**Pull request by meji46**
**Changes proposed:**

-  Implemented spell effect 286 (SPELL_EFFECT_GRANT_BATTLEPET_EXPERIENCE).
-  Implemented CriteriaType::BattlePetReachLevel & CriteriaType::ActivelyEarnPetLevel.
-  Implemented aura 420 (SPELL_AURA_MOD_BATTLE_PET_XP_PCT).

**Issues addressed:**
None

**Tests performed:**
Builds, tested in-game.

**Known issues and TODO list:** (add/remove lines as needed)
None

### CODE REVIEW

File: src/server/game/BattlePets/BattlePetMgr.cpp [hunk 1]
```diff
@@ -677,6 +678,72 @@ void BattlePetMgr::ChangeBattlePetQuality(ObjectGuid guid, BattlePetBreedQuality
...
+    {
+        xp -= nextLevelXp;
+
+        xpEntry = sBattlePetXPGameTable.GetRow(++level);
+        if (!xpEntry)
+            return;
+
+        nextLevelXp = uint16(GetBattlePetXPPerLevel(xpEntry));
+
+        player->UpdateCriteria(CriteriaType::BattlePetReachLevel, pet->PacketInfo.Species, level);
```
Comment (Shauren) [inline src/server/game/BattlePets/BattlePetMgr.cpp:720]: Why level is a miscValue2?

File: src/server/game/BattlePets/BattlePetMgr.cpp [hunk 2]
```diff
@@ -677,6 +678,72 @@ void BattlePetMgr::ChangeBattlePetQuality(ObjectGuid guid, BattlePetBreedQuality
     // _owner->GetPlayer()->SetCurrentBattlePetBreedQuality(qualityValue);
 }
 
+void BattlePetMgr::GrantBattlePetExperience(ObjectGuid guid, uint16 xp)
```
Comment (Shauren) [inline src/server/game/BattlePets/BattlePetMgr.cpp:681]: This should have another argument `xpSource` that will let you determine if aura bonus should be applied or if CriteriaType::ActivelyEarnPetLevel should be updated (level boosts shouldnt)

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/BattlePets/BattlePetMgr.cpp (hunk 1 above)
Comment (meji46) [reply src/server/game/BattlePets/BattlePetMgr.cpp:720]: In order to satisfy certain ModifierTreeType that use miscValue1 as a battle pet species (e.g. https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Achievements/CriteriaHandler.cpp#L2077)

File: src/server/game/BattlePets/BattlePetMgr.cpp (hunk 2 above)
Comment (meji46) [reply src/server/game/BattlePets/BattlePetMgr.cpp]: Added the new argument (enum class BattlePetXpSource)

**Review by Shauren — APPROVED** [review]

**Pull request merged**